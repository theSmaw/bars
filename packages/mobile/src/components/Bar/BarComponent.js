import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components/native/index';

export default function BarComponent({ limit, value }) {
  return (
    <Bar>
      <Progress
        limit={limit}
        value={value}
      >
        <Value>{value}</Value>
      </Progress>
    </Bar>
  );
}

BarComponent.propTypes = {
  limit: PropTypes.number.isRequired,
  value: PropTypes.number.isRequired
};

const Bar = styled.View`
  width: 500px;
`;

// In React-Native it's not possible to use mount from Enzyme. This makes it hard to test logic in
// styled-components using Jest Snapshots. I remain unconvinced that styled-components is the right
// solution.
/* istanbul ignore next */
const Progress = styled.View`
  display: flex;
  height: 35px;
  flex-direction: column;
  justify-content: center;
  margin: 0 0 20px 0;
  background-color: #66c1d2;
  width: ${props => (props.value / props.limit) * 100}%;
`;

const Value = styled.Text`
  color: #fff;
  text-align: center;
`;
