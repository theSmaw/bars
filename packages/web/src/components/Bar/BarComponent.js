import React from 'react';
import PropTypes from 'prop-types';
import styled from 'styled-components';

export default function BarComponent({ limit, value }) {
  return (
    <div className="bar">
      <Progress
        className="progress"
        role="progressbar"
        aria-valuenow={value}
        aria-valuemin="0"
        aria-valuemax={limit}
        limit={limit}
        value={value}
      >
        {value}
      </Progress>
    </div>
  );
}

BarComponent.propTypes = {
  limit: PropTypes.number.isRequired,
  value: PropTypes.number.isRequired
};

function getProgressWidth({ value, limit }) {
  return Math.min((value / limit) * 100, 100);
}

const Progress = styled.div`
  color: #fff;
  display: flex;
  height: 50px;
  flex-direction: column;
  justify-content: center;
  margin: 0 0 20px 0;
  text-align: center;
  background-color: ${props => (props.value > props.limit ? '#ff0000' : '#66c1d2')};
  transition: width .6s ease;
  width: ${getProgressWidth}%;
`;
