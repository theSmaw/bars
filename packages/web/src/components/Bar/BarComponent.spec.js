import React from 'react';
import { mount } from 'enzyme';

import BarComponent from './BarComponent';

describe('BarComponent', () => {
  it('Should render with the correct value', () => {
    const barComponent = mount(
      <BarComponent
        limit={50}
        value={5}
      />
    );

    expect(barComponent).toMatchSnapshot();
  });

  it('Should render values above the limit', () => {
    const barComponent = mount(
      <BarComponent
        limit={50}
        value={55}
      />
    );

    expect(barComponent).toMatchSnapshot();
  });
});
