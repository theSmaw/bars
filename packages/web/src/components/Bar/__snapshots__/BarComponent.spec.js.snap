// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BarComponent Should render values above the limit 1`] = `
<BarComponent
  limit={50}
  value={55}
>
  <div
    className="bar"
  >
    <styled.div
      aria-valuemax={50}
      aria-valuemin="0"
      aria-valuenow={55}
      className="progress"
      limit={50}
      role="progressbar"
      value={55}
    >
      <div
        aria-valuemax={50}
        aria-valuemin="0"
        aria-valuenow={55}
        className="progress sc-bdVaJa ccmbpX"
        role="progressbar"
        value={55}
      >
        55
      </div>
    </styled.div>
  </div>
</BarComponent>
`;

exports[`BarComponent Should render with the correct value 1`] = `
<BarComponent
  limit={50}
  value={5}
>
  <div
    className="bar"
  >
    <styled.div
      aria-valuemax={50}
      aria-valuemin="0"
      aria-valuenow={5}
      className="progress"
      limit={50}
      role="progressbar"
      value={5}
    >
      <div
        aria-valuemax={50}
        aria-valuemin="0"
        aria-valuenow={5}
        className="progress sc-bdVaJa kMmImj"
        role="progressbar"
        value={5}
      >
        5
      </div>
    </styled.div>
  </div>
</BarComponent>
`;
